### PR TITLE
added sigterm telemetry event api

### PR DESCRIPTION
**Pull request by gireesh-devtron**
# Description

added an api endpoint in telemetry client to send sigterm event to posthog.
This endpoint is used in prestop hook in orchestrator container to send sigterm event before stoping the container.

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

-> This api endpoint has been tested locally and able to send sigterm event from demo1 client.
image
-> able to send payload 
image

- [x] Test case A
- [ ] Test case B

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [x] I have commented my code, particularly in hard-to-understand areas
* [ ] I have tested it for all user roles
* [ ] I have added all the required unit/api test cases

### CODE REVIEW

**Comment by gireesh-devtron** [comment]
@manish-agrawal-ai please review this PR

File: pkg/user/UserAuthService.go [hunk 1]
```diff
@@ -366,6 +366,7 @@ func WhitelistChecker(url string) bool {
 		"/orchestrator/dashboard-event/dashboardLoggedIn",
 		"/orchestrator/self-register/check",
 		"/orchestrator/self-register",
+		"/orchestrator/telemetry/sigtermEvent",
```
Comment (manish-agrawal-ai) [inline pkg/user/UserAuthService.go:369]: unused import

File: pkg/gitops/GitOpsConfigService.go [hunk 2]
```diff
@@ -126,6 +126,15 @@ func NewGitOpsConfigServiceImpl(Logger *zap.SugaredLogger, ciHandler pipeline.Ci
 
 func (impl *GitOpsConfigServiceImpl) ValidateAndCreateGitOpsConfig(config *bean2.GitOpsConfigDto) (DetailedErrorGitOpsConfigResponse, error) {
 	detailedErrorGitOpsConfigResponse := impl.GitOpsValidateDryRun(config)
+	//hostUrl := config.Host
```
Comment (manish-agrawal-ai) [inline pkg/gitops/GitOpsConfigService.go:129]: remove unused code

Comment (gireesh-devtron) [reply pkg/gitops/GitOpsConfigService.go:129]: this was not written by me ,this is merged from devtron:main

File: api/router/TelemetryRouter.go [hunk 3]
```diff
@@ -44,5 +44,7 @@ func (router TelemetryRouterImpl) InitTelemetryRouter(telemetryRouter *mux.Route
 		HandlerFunc(router.handler.GetTelemetryMetaInfo).Methods("GET")
 	telemetryRouter.Path("/event").
 		HandlerFunc(router.handler.SendTelemetryData).Methods("POST")
+	telemetryRouter.Path("/sigtermEvent").
+		HandlerFunc(router.handler.SigtermEventHandler).Methods("GET")
```
Comment (manish-agrawal-ai) [inline api/router/TelemetryRouter.go:48]: method should be POST

Comment (gireesh-devtron) [reply api/router/TelemetryRouter.go:48]: the hook only supports get method

File: api/router/TelemetryRouter.go [hunk 4]
```diff
@@ -44,5 +44,7 @@ func (router TelemetryRouterImpl) InitTelemetryRouter(telemetryRouter *mux.Route
 		HandlerFunc(router.handler.GetTelemetryMetaInfo).Methods("GET")
 	telemetryRouter.Path("/event").
 		HandlerFunc(router.handler.SendTelemetryData).Methods("POST")
+	telemetryRouter.Path("/sigtermEvent").
```
Comment (manish-agrawal-ai) [inline api/router/TelemetryRouter.go:47]: please change to /summary

File: api/restHandler/TelemetryRestHandler.go [hunk 5]
```diff
@@ -92,3 +93,8 @@ func (handler TelemetryRestHandlerImpl) SendTelemetryData(w http.ResponseWriter,
 	common.WriteJsonResp(w, nil, "success", http.StatusOK)
 
 }
+
+func (handler TelemetryRestHandlerImpl) SigtermEventHandler(w http.ResponseWriter, r *http.Request) {
+	handler.telemetryEventClient.SendSigtermSummaryEvent()
```
Comment (manish-agrawal-ai) [inline api/restHandler/TelemetryRestHandler.go:98]: Read eventType from request. (You should send as "SIG_TERM" from hook. )
If eventType is nil or empty then default eventType should be Summary.

File: client/telemetry/TelemetryEventClient.go [hunk 6]
```diff
@@ -46,6 +46,7 @@ type TelemetryEventClient interface {
 	SendTelemetryDashboardAccessEvent() error
 	SendTelemetryDashboardLoggedInEvent() error
 	SendGenericTelemetryEvent(eventType string, prop map[string]interface{}) error
+	SendSigtermSummaryEvent()
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClient.go:49]: SendSummaryEvent

File: client/telemetry/TelemetryEventClient.go [hunk 7]
```diff
@@ -165,10 +167,20 @@ func (impl *TelemetryEventClientImpl) SummaryDetailsForTelemetry() (cluster []cl
...
+	var eventType = Sigterm
+	impl.ArgumentedSummaryEventForTelemetryEA(eventType)
+}
+
 func (impl *TelemetryEventClientImpl) SummaryEventForTelemetryEA() {
+	var eventType = Summary
+	impl.ArgumentedSummaryEventForTelemetryEA(eventType)
+}
+
+func (impl *TelemetryEventClientImpl) ArgumentedSummaryEventForTelemetryEA(eventType TelemetryEventType) {
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClient.go:180]: need refactoring. (better naming)

File: api/restHandler/TelemetryRestHandler.go [hunk 8]
```diff
@@ -92,3 +95,27 @@ func (handler TelemetryRestHandlerImpl) SendTelemetryData(w http.ResponseWriter,
...
+	err := decoder.Decode(&payload)
+	if err != nil && err != io.EOF {
+		handler.logger.Errorw("request err, SendsummaryEvent", "err", err, "payload", payload)
+		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
+		return
+	}
+
+	var eventType string
+	if err == io.EOF {
+		eventType = "Summary"
```
Comment (manish-agrawal-ai) [inline api/restHandler/TelemetryRestHandler.go:111]: Read "Summary" from constant

File: api/restHandler/TelemetryRestHandler.go [hunk 9]
```diff
@@ -92,3 +95,27 @@ func (handler TelemetryRestHandlerImpl) SendTelemetryData(w http.ResponseWriter,
...
+	} else {
+		eventType = payload["eventType"].(string)
+		if eventType == "" {
+			eventType = "Summary"
```
Comment (manish-agrawal-ai) [inline api/restHandler/TelemetryRestHandler.go:115]: Read "Summary" from constant

File: api/restHandler/TelemetryRestHandler.go [hunk 10]
```diff
@@ -92,3 +95,27 @@ func (handler TelemetryRestHandlerImpl) SendTelemetryData(w http.ResponseWriter,
...
+		}
+	}
+	fmt.Println(eventType)
```
Comment (manish-agrawal-ai) [inline api/restHandler/TelemetryRestHandler.go:118]: remove print

File: api/restHandler/TelemetryRestHandler.go [hunk 11]
```diff
@@ -92,3 +95,27 @@ func (handler TelemetryRestHandlerImpl) SendTelemetryData(w http.ResponseWriter,
 	common.WriteJsonResp(w, nil, "success", http.StatusOK)
 
 }
+
+func (handler TelemetryRestHandlerImpl) SendSummaryEvent(w http.ResponseWriter, r *http.Request) {
```
Comment (manish-agrawal-ai) [inline api/restHandler/TelemetryRestHandler.go:99]: add a logging here

File: client/telemetry/TelemetryEventClient.go [hunk 12]
```diff
@@ -122,6 +123,7 @@ const (
 	InstallationApplicationError TelemetryEventType = "InstallationApplicationError"
 	DashboardAccessed            TelemetryEventType = "DashboardAccessed"
 	DashboardLoggedIn            TelemetryEventType = "DashboardLoggedIn"
+	Sigterm                      TelemetryEventType = "Sigterm"
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClient.go:126]: remove this as I think this is not in use

File: client/telemetry/TelemetryEventClient.go [hunk 13]
```diff
@@ -166,9 +168,13 @@ func (impl *TelemetryEventClientImpl) SummaryDetailsForTelemetry() (cluster []cl
 }
 
 func (impl *TelemetryEventClientImpl) SummaryEventForTelemetryEA() {
+	impl.SendSummaryEvent("Summary")
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClient.go:171]: Read "Summary" from constant (TelemetryEventType Summary)

File: client/telemetry/TelemetryEventClient.go [hunk 14]
```diff
@@ -166,9 +168,13 @@ func (impl *TelemetryEventClientImpl) SummaryDetailsForTelemetry() (cluster []cl
...
+}
+
+func (impl *TelemetryEventClientImpl) SendSummaryEvent(eventType string) {
 	ucid, err := impl.getUCID()
 	if err != nil {
-		impl.logger.Errorw("exception caught inside telemetry summary event", "err", err)
+		impl.logger.Errorw("exception caught inside telemetry summary for"+eventType+"event", "err", err)
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClient.go:177]: impl.logger.Errorw("exception caught inside telemetry summary event", "eventType", "eventType", "err", err)

File: client/telemetry/TelemetryEventClient.go [hunk 15]
```diff
@@ -166,9 +168,13 @@ func (impl *TelemetryEventClientImpl) SummaryDetailsForTelemetry() (cluster []cl
 }
 
 func (impl *TelemetryEventClientImpl) SummaryEventForTelemetryEA() {
+	impl.SendSummaryEvent("Summary")
+}
+
+func (impl *TelemetryEventClientImpl) SendSummaryEvent(eventType string) {
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClient.go:174]: add a logger here and add eventType in that

File: client/telemetry/TelemetryEventClient.go [hunk 16]
```diff
@@ -189,19 +195,19 @@ func (impl *TelemetryEventClientImpl) SummaryEventForTelemetryEA() {
 
 	reqBody, err := json.Marshal(payload)
 	if err != nil {
-		impl.logger.Errorw("SummaryEventForTelemetry, payload marshal error", "error", err)
+		impl.logger.Errorw("SummaryEventForTelemetry in", eventType+"event", "payload marshal error", "error", err)
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClient.go:198]: Correct log everywhere

File: client/telemetry/TelemetryEventClientExtended.go [hunk 17]
```diff
@@ -97,39 +97,43 @@ func NewTelemetryEventClientImplExtended(logger *zap.SugaredLogger, client *http
...
+	Build                                bool               `json:"build,omitempty"`
+	Deployment                           bool               `json:"deployment,omitempty"`
+	ServerVersion                        string             `json:"serverVersion,omitempty"`
+	DevtronGitVersion                    string             `json:"devtronGitVersion,omitempty"`
+	DevtronVersion                       string             `json:"devtronVersion,omitempty"`
+	DevtronMode                          string             `json:"devtronMode,omitempty"`
 }
 
 func (impl *TelemetryEventClientImplExtended) SummaryEventForTelemetry() {
+	impl.SendSummaryEvent("Summary")
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClientExtended.go:133]: Read "Summary" from constant

File: client/telemetry/TelemetryEventClientExtended.go [hunk 18]
```diff
@@ -97,39 +97,43 @@ func NewTelemetryEventClientImplExtended(logger *zap.SugaredLogger, client *http
...
+}
+
+func (impl *TelemetryEventClientImplExtended) SendSummaryEvent(eventType string) {
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClientExtended.go:136]: add logger

**Review by manish-agrawal-ai — COMMENTED** [review]
check comments

File: api/restHandler/TelemetryRestHandler.go [hunk 19]
```diff
@@ -44,6 +46,10 @@ type TelemetryGenericEvent struct {
 	eventPayload map[string]interface{}
 }
 
+const (
+	summary string = "Summary"
```
Comment (manish-agrawal-ai) [inline api/restHandler/TelemetryRestHandler.go:50]: TelemetryEventTypeSummary

File: api/restHandler/TelemetryRestHandler.go (hunk 11 above)
Comment (manish-agrawal-ai) [reply api/restHandler/TelemetryRestHandler.go:102]: impl.logger.Info("SendSummaryEvent came")

File: client/telemetry/TelemetryEventClient.go [hunk 20]
```diff
@@ -166,20 +167,27 @@ func (impl *TelemetryEventClientImpl) SummaryDetailsForTelemetry() (cluster []cl
 }
 
 func (impl *TelemetryEventClientImpl) SummaryEventForTelemetryEA() {
+	err := impl.SendSummaryEvent(string(Summary))
+	if err != nil {
+		impl.logger.Errorw("error occurred in SummaryEventForTelemetryEA", "err", err)
+	}
+}
+
+func (impl *TelemetryEventClientImpl) SendSummaryEvent(eventType string) error {
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClient.go:176]: impl.logger.Infow("Sending summary event", "eventType", eventType)

File: api/restHandler/TelemetryRestHandler.go [hunk 21]
```diff
@@ -92,3 +94,33 @@ func (handler TelemetryRestHandlerImpl) SendTelemetryData(w http.ResponseWriter,
 	common.WriteJsonResp(w, nil, "success", http.StatusOK)
 
 }
+
+func (handler TelemetryRestHandlerImpl) SendSummaryEvent(w http.ResponseWriter, r *http.Request) {
+	handler.logger.Infow("Handling SendSummaryEvent request")
```
Comment (manish-agrawal-ai) [inline api/restHandler/TelemetryRestHandler.go:99]: remove "w". it should be handler.logger.Info

File: client/telemetry/TelemetryEventClient.go [hunk 22]
```diff
@@ -166,20 +168,28 @@ func (impl *TelemetryEventClientImpl) SummaryDetailsForTelemetry() (cluster []cl
 }
 
 func (impl *TelemetryEventClientImpl) SummaryEventForTelemetryEA() {
+	err := impl.SendSummaryEvent(string(Summary))
+	if err != nil {
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClient.go:172]: these err change is not required

File: client/telemetry/TelemetryEventClient.go [hunk 23]
```diff
@@ -166,20 +168,28 @@ func (impl *TelemetryEventClientImpl) SummaryDetailsForTelemetry() (cluster []cl
...
+	}
+}
+
+func (impl *TelemetryEventClientImpl) SendSummaryEvent(eventType string) error {
+	impl.logger.Infow("sending summary event", "eventType", eventType)
 	ucid, err := impl.getUCID()
 	if err != nil {
-		impl.logger.Errorw("exception caught inside telemetry summary event", "err", err)
-		return
+		impl.logger.Errorw("exception caught inside SendSummaryEvent ", "eventType", eventType, "err", err)
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClient.go:181]: there is no need to change loggging

File: client/telemetry/TelemetryEventClient.go [hunk 24]
```diff
@@ -190,19 +200,20 @@ func (impl *TelemetryEventClientImpl) SummaryEventForTelemetryEA() {
...
-		return
+		return err
 	}
 
-	err = impl.EnqueuePostHog(ucid, Summary, prop)
+	err = impl.EnqueuePostHog(ucid, TelemetryEventType(eventType), prop)
 	if err != nil {
 		impl.logger.Errorw("SummaryEventForTelemetry, failed to push event", "ucid", ucid, "error", err)
 	}
+	return nil
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClient.go:216]: return err

File: client/telemetry/TelemetryEventClientExtended.go [hunk 25]
```diff
@@ -255,17 +263,18 @@ func (impl *TelemetryEventClientImplExtended) SummaryEventForTelemetry() {
...
 	if err != nil {
 		impl.logger.Errorw("SummaryEventForTelemetry, payload unmarshal error", "error", err)
-		return
+		return err
 	}
 
-	err = impl.EnqueuePostHog(ucid, Summary, prop)
+	err = impl.EnqueuePostHog(ucid, TelemetryEventType(eventType), prop)
 	if err != nil {
 		impl.logger.Errorw("SummaryEventForTelemetry, failed to push event", "ucid", ucid, "error", err)
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClientExtended.go:277]: return err just below this

File: client/telemetry/TelemetryEventClient.go [hunk 26]
```diff
@@ -190,19 +200,20 @@ func (impl *TelemetryEventClientImpl) SummaryEventForTelemetryEA() {
...
 	if err != nil {
 		impl.logger.Errorw("SummaryEventForTelemetry, payload unmarshal error", "error", err)
-		return
+		return err
 	}
 
-	err = impl.EnqueuePostHog(ucid, Summary, prop)
+	err = impl.EnqueuePostHog(ucid, TelemetryEventType(eventType), prop)
 	if err != nil {
 		impl.logger.Errorw("SummaryEventForTelemetry, failed to push event", "ucid", ucid, "error", err)
```
Comment (manish-agrawal-ai) [inline client/telemetry/TelemetryEventClient.go:214]: return err just below this

**Review by manish-agrawal-ai — DISMISSED** [review]

File: cmd/external-app/externalApp.go [hunk 27]
```diff
@@ -60,5 +63,10 @@ func (app *App) Start() {
 }
 
 func (app *App) Stop() {
-
+	app.Logger.Info("orchestrator shutdown initiating")
+	posthogCl := app.posthogClient.Client
+	if posthogCl != nil {
+		app.Logger.Infow("flushing messages of posthog")
```
Comment (manish-agrawal-ai) [inline cmd/external-app/externalApp.go:69]: remove "W"

File: App.go [hunk 28]
```diff
@@ -111,6 +115,11 @@ func (app *App) Start() {
 
 func (app *App) Stop() {
 	app.Logger.Infow("orchestrator shutdown initiating")
+	posthogCl := app.posthogClient.Client
+	if posthogCl != nil {
+		app.Logger.Infow("flushing messages of posthog")
```
Comment (manish-agrawal-ai) [inline App.go:120]: remove "W"

File: DockerfileEA [hunk 29]
```diff
@@ -4,6 +4,7 @@ RUN echo $GOPATH
 
 RUN apk add --no-cache git gcc musl-dev
 RUN apk add --update make
+RUN apk add curl
```
Comment (manish-agrawal-ai) [inline DockerfileEA:7]: we pushed this change to below?

**Review by manish-agrawal-ai — COMMENTED** [review]
manifest change is remaining

**Review by manish-agrawal-ai — APPROVED** [review]
approved

**Review by pawan-59 — COMMENTED** [review]
Please include the lifecycle configuration in charts/devtron/templates/devtron.yaml too

**Review by manish-agrawal-ai — APPROVED** [review]

**Pull request merged**